zerver/lib/narrow.py

```diff
@@ -861,6 +873,65 @@ def get_channel_from_narrow_access_unchecked(
     return None
 
 
+# This function corrects the conversation filters of the narrow - channel,
+# topic, DM conversation to point to the narrow in which the message
+# id operand of the `with` operator is present in.
+#
+# This is done as an effort to facilitate permalinks for topics -- if there
+# exists a message represented by the id (`with` operand), we redirect to
+# the narrow in which the message is present. Else, we fallback
+# to the narrow mentioned originally.
+def update_narrow_terms_containing_with_operator(
+    realm: Realm,
+    maybe_user_profile: Union[UserProfile, AnonymousUser],
+    narrow: Optional[List[NarrowParameter]],
+) -> Optional[List[NarrowParameter]]:
+    if narrow is None:
+        return narrow
+
+    contains_with_operator = list(filter(lambda term: term.operator == "with", narrow))
+
+    if len(contains_with_operator) > 0:
+        with_term = contains_with_operator[0]
+        with_term_index = narrow.index(with_term)
+        message_id = int(with_term.operand)
+
+        narrow.remove(with_term)
+
+        if maybe_user_profile.is_authenticated:
+            try:
+                message = access_message(maybe_user_profile, message_id)
+            except JsonableError:
+                return narrow
+        else:
+            try:
+                message = access_web_public_message(realm, message_id)
+            except MissingAuthenticationError:
+                return narrow
+
+        narrow = narrow[with_term_index:]
+
+        if message.is_stream_message():
+            channel_id = message.recipient.type_id
+            topic = message.topic_name()
+            channel_topic_narrow = [
+                NarrowParameter(operator="channel", operand=channel_id),
+                NarrowParameter(operator="topic", operand=topic),
+            ]
+            web_public_narrow = [NarrowParameter(operator="channels", operand="web-public")]
+            narrow = channel_topic_narrow + narrow
+
+            if not maybe_user_profile.is_authenticated:
+                narrow = web_public_narrow + narrow
+
+        else:
+            huddle_user_ids = list(get_huddle_user_ids(message.recipient))
+            dm_narrow = [NarrowParameter(operator="dm", operand=huddle_user_ids)]
+            narrow = dm_narrow + narrow
```

Comment:
This seems to assume that the only possibilities are streams and Huddles, but 1:1 group DMs are represented differently; does this work correctly?

Reply:
Considering [this](https://github.com/zulip/zulip/pull/30114#discussion_r1611732823) comment, is it right to just remove the `with` term in case recipient type is not a stream?

Reply:
updated!